apps/rush-lib/src/cli/utilities/ShrinkwrapFile.ts

```diff
@@ -128,18 +57,23 @@ export default class ShrinkwrapFile {
     }
   }
 
-  private constructor(shrinkwrapJson: IShrinkwrapJson) {
-    this._shrinkwrapJson = shrinkwrapJson;
+  /**
+   * Returns the list of temp projects defined in this file.
+   * Example: [ '@rush-temp/project1', '@rush-temp/project2' ]
+   */
+  public abstract getTempProjectNames(): ReadonlyArray<string>;
+
+  protected abstract getDependencyVersion(dependencyName: string, tempProjectName?: string): string | undefined;
 
-    // Normalize the data
-    if (!this._shrinkwrapJson.version) {
-      this._shrinkwrapJson.version = '';
-    }
-    if (!this._shrinkwrapJson.name) {
-      this._shrinkwrapJson.name = '';
-    }
-    if (!this._shrinkwrapJson.dependencies) {
-      this._shrinkwrapJson.dependencies = { };
+  protected _getTempProjectNames(dependencies: { [key: string]: {} } ): ReadonlyArray<string> {
```

Comment:
>_getTempProjectNames [](start = 12, length = 20)

If a method is exactly the same for NPM and PNPM, then I'm fine with putting in the base class.

If parts of the method are mostly the same, but random parts are different, distending the logic across child methods that rely on base class protected members with the same name -- that requires the reader to yo-yo up and down the class hierarchy to understand the business logic.  Given that NPM SUPPORT IS GOING TO BE DELETED AS SOON AS POSSIBLE, I'd rather just have two copies of the code that I can understand and test independently.

For example, if NPM is already "working", why do I have to retest it every time I fix a bug in PNPM?

Reply:
Advice taken

---
In reply to: [151790087](https://github.com/Microsoft/web-build-tools/pull/427#discussion_r151790087) [](ancestors = 151790087)